### PR TITLE
[Core] Fix GPU first scheduling that is not working with placement group

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This is one of implementation suggestion. I discovered the root cause, so we can discuss if my implementation makes sense (if we have a better option, we can choose that option).

cc @sasha-s @ericl @scv119 please share you guys thoughts. 

https://github.com/ray-project/ray/pull/19129 disables the GPU scheduling policy due to instability in placement group.

### Hybrid Policy API
When we pick the best node with the hybrid policy with /*require_available = false*/, we have an invariant; 
- The function will return the best node if any node is "feasible" to schedule.

Basically the function is like this;
- Find the best node that's available
- Find the best node that's feasible

### GPU scheduling Logic
This is how we chooses a node when gpu scheduling is enabled.
1. Find the best node with non-GPU nodes
2. If it cannot find the best node, schedule with GPU nodes.
Ref: https://github.com/ray-project/ray/pull/18615/files#diff-8a271b85b61542daedc5dd81fd8922b15634535871ecac7d37e99b6c80058536R61 

## Problem
Imagine this scenario
2 nodes; GPU node (3 CPUs), and non-GPU node (3 CPUs)
Placement group with {1 GPU, 1 CPU}  and {1 CPU} * 5

The placement group scheduling hangs because

```
Imagine the non-GPU node CPU bundles are all occupied.
1. non-GPU node has no more pg resources.
2. We should spillback the task to GPU node now.
3. Scheduling policy starts
4. We chooses the best node from non-GPU nodes first. 
5. The policy returns itself as a "feasible" best node. 
6. The function returns the non-GPU node, that's already full of resource usage because it is feasible.
7. Hangs forever because the non-GPU node uses actors
```

## Solution
The solution is simple. We try scheduling on CPU only nodes first, but with /*require_available=true*/ and do the same thing to. the gpu node. This will make sure we can spillback tasks if they are not "available" and only "feasible".

And if both cannot find the best node, we fallback to the original policy, which guarantees the liveness. 

I verified it works with the repro. 

## Related issue number

https://github.com/ray-project/ray/issues/19130

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by sasha-s** [comment]
Looks reasonable to me.
I actually thought about doing it in the first place, but it looked like an overkill at a time.

File: src/ray/raylet/scheduling/scheduling_policy.cc [hunk 1]
```diff
@@ -127,6 +127,7 @@ int64_t HybridPolicyWithFilter(const ResourceRequest &resource_request,
                critical_resource_utilization < best_utilization_score &&
                !require_available) {
       // Pick the best feasible node by critical resource utilization.
+      RAY_LOG(INFO) << "Best is not available and critical utilization is high enough";
```
Comment (ericl) [inline src/ray/raylet/scheduling/scheduling_policy.cc:130]: Remove this

**Review by ericl — COMMENTED** [review]
The solution makes sense to me. How can we test this?

**Comment by ericl** [comment]
I see, can we flip on the flag and check the test goes from failing to passing?

**Comment by scv119** [comment]
Looks like a reasonable fix.

**Review by sasha-s — APPROVED** [review]
LGTM

**Comment by rkooo567** [comment]
Windows failure: I will skip tests in windows as every other test in test scheduling has been disabled there

**Pull request merged**